src/main/kotlin/org/jitsi/nlj/transform/node/AudioRedHandler.kt

```diff
@@ -0,0 +1,275 @@
+
+            sentAudioCache.insert(packetInfo.packetAs())
+
+            if (encapsulate) {
+                val redundancy = mutableListOf<RtpPacket>()
+                val seq = audioRtpPacket.sequenceNumber
+
+                when (config.distance) {
+                    RedDistance.ONE -> {
+                        getPacketToProtect(applySequenceNumberDelta(seq, -1), config.vadOnly)?.also { secondary ->
+                            redundancy.add(secondary)
+                            stats.redundancyPacketAdded()
+                        }
+                    }
+                    RedDistance.TWO -> {
+                        getPacketToProtect(applySequenceNumberDelta(seq, -1), false)?.also { secondary ->
+                            // With distance 2 we only add the tertiary packet when there is a secondary available
+                            // (regardless of secondary's VAD). This guarantees that the sequence numbers of the
+                            // redundancy packets always directly proceed the primary packet, i.e. that we don't encode
+                            // a packet with primary seq=N and a single redundancy with seq=N-2. This is be important
+                            // when the receiver of the RED stream is another jitsi-videobridge instance (via Octo),
+                            // which makes that assumption about the stream it receives.
+                            val tertiary = getPacketToProtect(applySequenceNumberDelta(seq, -2), config.vadOnly)
+                            if (tertiary != null) {
+                                redundancy.add(tertiary)
+                                stats.redundancyPacketAdded()
+                                redundancy.add(secondary)
+                                stats.redundancyPacketAdded()
+                            } else if (!config.vadOnly || secondary.hasVad()) {
+                                // If there's no tertiary encode the secondary alone, but this time check its VAD.
+                                redundancy.add(secondary)
+                                stats.redundancyPacketAdded()
+                            }
+                        }
+                    }
+                }
+
+                val redPacket = RedAudioRtpPacket.builder.build(redPayloadType!!, audioRtpPacket, redundancy)
+                packetInfo.packet = redPacket
+
+                redundancy.forEach { BufferPool.returnBuffer(it.buffer) }
+                BufferPool.returnBuffer(audioRtpPacket.buffer)
+
+                stats.audioPacketEncapsulated()
+            } else {
+                stats.audioPacketForwarded()
+            }
+            return packetInfo
+        }
+
+        private fun getPacketToProtect(seq: Int, vadOnly: Boolean): RtpPacket? {
+            // All of the transform pipeline runs in a single thread, and we only use the packet momentarily to make a
+            // copy into a new RED packet, so its safe to just peek() at it.
```

Comment:
If you know you never want to clone for this object, you can pass a no-op function as the cache's `clone` method.  (This is what I do for a few of my uses of ArrayCache, e.g. `ResumableStreamRewriter`.)

Reply:
I want to clone on insertion (since otherwise the buffers will be returned), but not on retrieval (since I know it's not necessary).